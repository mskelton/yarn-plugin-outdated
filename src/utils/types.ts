export interface DependencyInfo {
  name: string
  current: string
  latest: string
  type: "dependencies" | "devDependencies"
}
