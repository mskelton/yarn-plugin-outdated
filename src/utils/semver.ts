import * as semver from "semver"

export const parseVersion = (version: string) =>
  semver.valid(semver.coerce(version))!
