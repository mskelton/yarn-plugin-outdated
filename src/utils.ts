import { Manifest } from "@yarnpkg/core"
import { fromUrl } from "hosted-git-info"
import semver from "semver"

export const truthy = Boolean as unknown as <T>(
  arg: T | undefined | null | false
) => arg is T

export function getHomepageURL({ raw: manifest }: Manifest): string | null {
  const repo = manifest.repository
  const repoURL = manifest.homepage
    ? manifest.homepage
    : typeof repo === "string"
    ? repo
    : typeof repo === "object" && typeof repo.url === "string"
    ? repo.url
    : null

  const info = repoURL ? fromUrl(repoURL) : undefined
  const commitish = info?.committish ? `#${info.committish}` : ""

  return info
    ? `https://${info.domain}/${info.user}/${info.project}${commitish}`
    : repoURL
}

/**
 * Because some packages have a pre-release version as their `latest` version,
 * we need to first check if the latest version is a pre-release. If it is,
 * we compare the current and latest directly, otherwise we coerce the current
 * version to remove any pre-release identifiers to determine if it is outdated.
 */
export function isVersionOutdated(current: string, latest: string) {
  return semver.parse(latest)!.prerelease.length
    ? semver.lt(current, latest)
    : semver.lt(semver.coerce(current)!, latest)
}
