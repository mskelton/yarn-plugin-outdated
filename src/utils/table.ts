import { Configuration, formatUtils } from "@yarnpkg/core"
import * as semver from "semver"
import { DependencyInfo } from "./types"

const columns = ["name", "current", "latest", "type"] as const

export type TableColumn = typeof columns[number]

export const headers: Record<TableColumn, string> = {
  current: "Current",
  latest: "Latest",
  name: "Package",
  type: "Package Type",
}

// Iterates through the dependencies to determine the necessary size
// of each column.
function getColumnSizes(
  dependencies: DependencyInfo[]
): Record<TableColumn, number> {
  const sizes = {
    current: headers.current.length,
    latest: headers.latest.length,
    name: headers.name.length,
    type: headers.type.length,
  }

  for (const dependency of dependencies) {
    for (const [key, value] of Object.entries(dependency)) {
      const acc = sizes[key as TableColumn]
      const cur = value.length

      sizes[key as TableColumn] = acc > cur ? acc : cur
    }
  }

  return sizes
}

function printHeader(
  configuration: Configuration,
  sizes: Record<TableColumn, number>
) {
  const format = (key: TableColumn) =>
    formatUtils.pretty(configuration, headers[key].padEnd(sizes[key]), "bold")

  printRow({
    current: format("current"),
    latest: format("latest"),
    name: format("name"),
    type: format("type"),
  })
}

function printRow(row: Record<TableColumn, string>) {
  console.log(columns.map((column) => row[column]).join("   "))
}

export function printTable(
  configuration: Configuration,
  dependencies: DependencyInfo[]
) {
  const sizes = getColumnSizes(dependencies)
  printHeader(configuration, sizes)

  dependencies.forEach((dep) => {
    const current = semver.parse(dep.current)!
    const latest = semver.parse(dep.latest)!

    const color =
      latest.major > current.major
        ? "red"
        : latest.minor > current.minor
        ? "yellow"
        : "green"

    printRow({
      current: dep.current.padEnd(sizes.current),
      latest: dep.latest.padEnd(sizes.latest),
      name: formatUtils.pretty(
        configuration,
        dep.name.padEnd(sizes.name),
        color
      ),
      type: dep.type.padEnd(sizes.type),
    })
  })
}
