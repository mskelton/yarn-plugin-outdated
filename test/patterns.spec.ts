import { makeTemporaryEnv } from "./utils/env"

const manifest = {
  dependencies: {
    "@scoped/minor": "1.0.0",
    "@scoped/patch": "1.0.0",
    minor: "1.0.0",
    patch: "1.0.0",
  },
}

it("can filter by an exact package name", async () => {
  const { run, writeJSON } = await makeTemporaryEnv()

  await writeJSON("package.json", manifest)
  await run("install")

  const { stderr, stdout } = await run("outdated patch")
  expect(stdout).toMatchSnapshot()
  expect(stderr).toBe("")
})

it("can filter using a wildcard", async () => {
  const { run, writeJSON } = await makeTemporaryEnv()

  await writeJSON("package.json", manifest)
  await run("install")

  const { stderr, stdout } = await run("outdated @scoped/*")
  expect(stdout).toMatchSnapshot()
  expect(stderr).toBe("")
})

it("should throw an error if the pattern doesn't match any packages", async () => {
  const { run, writeJSON } = await makeTemporaryEnv()

  await writeJSON("package.json", manifest)
  await run("install")

  const { stderr, stdout } = await run("outdated not-a-package")
  expect(stdout).toMatchSnapshot()
  expect(stderr).toBe("")
})
