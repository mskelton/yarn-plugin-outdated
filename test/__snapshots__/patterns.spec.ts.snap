// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can filter by an exact package name 1`] = `
"➤ YN0000: ┌ Checking for outdated dependencies
➤ YN0000: └ Completed

➤ YN0000: Package   Current   Latest   Package Type
➤ YN0000: patch     1.0.0     1.0.1    dependencies

➤ YN0000: 1 dependency is out of date
➤ YN0000: Done with warnings
"
`;

exports[`can filter using a wildcard 1`] = `
"➤ YN0000: ┌ Checking for outdated dependencies
➤ YN0000: └ Completed

➤ YN0000: Package         Current   Latest   Package Type
➤ YN0000: @scoped/minor   1.0.0     1.1.0    dependencies
➤ YN0000: @scoped/patch   1.0.0     1.0.1    dependencies

➤ YN0000: 2 dependencies are out of date
➤ YN0000: Done with warnings
"
`;

exports[`should throw an error if the pattern doesn't match any packages 1`] = `
"Usage Error: Pattern not-a-package doesn't match any packages referenced by any workspace

$ yarn outdated [-a,--all] [-c,--check] [--json] ...
"
`;
